mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala

```diff
@@ -0,0 +1,690 @@
+    logInfo(s"Building the cluster tree is started in ${sc.appName}")
+    val root = BisectingKMeans.buildTree(nodes, BisectingKMeans.ROOT_INDEX_KEY, this.k)
+    if (root.isEmpty) {
+      new SparkException("Failed to build a cluster tree from a Map type of clusterStats")
+    }
+
+    // set the elapsed time for training
+    val finishTime = (System.currentTimeMillis() - startTime) / 1000.0
+    logInfo(s"Elapsed Time for ${this.getClass.getSimpleName} Training: ${finishTime} [sec]")
+
+    // make a bisecting kmeans model
+    val model = new BisectingKMeansModel(root.get)
+    val leavesNodes = model.getClusters
+    if (leavesNodes.length < this.k) {
+      logWarning(s"# clusters is less than you want: ${leavesNodes.length} / ${k}")
+    }
+    model
+  }
+}
+
+
+private[clustering] object BisectingKMeans {
+
+  val ROOT_INDEX_KEY: BigInt = 1
+
+  /**
+   * Finds the closes cluster's center
+   *
+   * @param metric a distance metric
+   * @param centers centers of the clusters
+   * @param point a target point
+   * @return an index of the array of clusters
+   */
+  def findClosestCenter(metric: (BV[Double], BV[Double]) => Double)
+      (centers: Seq[BV[Double]])(point: BV[Double]): Int = {
+    val (closestCenter, closestIndex) =
+      centers.zipWithIndex.map { case (center, idx) => (metric(center, point), idx)}.minBy(_._1)
+    closestIndex
+  }
+
+  /**
+   * Summarizes data by each cluster as Map
+   *
+   * @param data pairs of point and its cluster index
+   */
+  def summarizeClusters(data: RDD[(BigInt, BV[Double])]): Map[BigInt, BisectingClusterStat] = {
+
+    val stats = data.mapPartitions { iter =>
+      // calculate the accumulation of the all point in a partition and count the rows
+      val map = mutable.Map.empty[BigInt, (BV[Double], Double, BV[Double])]
+      iter.foreach { case (idx: BigInt, point: BV[Double]) =>
+        // get a map value or else get a sparse vector
+        val (sumBV, n, sumOfSquares) = map
+          .getOrElse(idx, (BSV.zeros[Double](point.size), 0.0, BSV.zeros[Double](point.size)))
+        map(idx) = (sumBV + point, n + 1.0, sumOfSquares + (point :* point))
+      }
+      map.toIterator
+    }.reduceByKey { case ((sum1, n1, sumOfSquares1), (sum2, n2, sumOfSquares2)) =>
+      // sum the accumulation and the count in the all partition
+      (sum1 + sum2, n1 + n2, sumOfSquares1 + sumOfSquares2)
+    }.collect().toMap
+
+    stats.map {case (i, stat) => i -> new BisectingClusterStat(stat._2.toLong, stat._1, stat._3)}
+  }
+
+  /**
+   * Assigns the initial cluster index id to all data
+   */
+  def initData(data: RDD[Vector]): RDD[(BigInt, BV[Double])] = {
+    data.map { v: Vector => (BisectingKMeans.ROOT_INDEX_KEY, v.toBreeze)}
+  }
+
+  /**
+   * Gets the initial centers for bisect k-means
+   *
+   * @param data pairs of point and its cluster index
+   * @param stats pairs of cluster index and cluster statistics
+   */
+  def initNextCenters(
```

Comment:
maybe we can do this:
1. select a random (with pre-determined seed) vector `p` of length in-cluster standard deviation for perturbation
2. use `c + p` and `c - p` as the new initial centers